File: all/234_palindrome_linked_list/sol_recursive.c
#include <stdbool.h>
#include <stddef.h>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

bool checkPalindromeRecursively(
    struct ListNode **left_node_ptr_ptr,
    const struct ListNode * const current_node_ptr
) {
    if (NULL == current_node_ptr) {
        return true;
    }

    if (
        !checkPalindromeRecursively(
            left_node_ptr_ptr,
            current_node_ptr->next
        )
    ) {
        return false;
    }

    bool current_result = (
        (current_node_ptr->val) == ((*left_node_ptr_ptr)->val)
    );

    (*left_node_ptr_ptr) = ((*left_node_ptr_ptr)->next);

    return current_result;
}

bool isPalindrome(struct ListNode* head) {
    struct ListNode * left_node_ptr = head;

    return checkPalindromeRecursively(&left_node_ptr, head);
}

File: all/234_palindrome_linked_list/sol_recursive.py
from typing import Optional


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        left_node = head

        def checkPalindromeRecursively(
            current_node: Optional[ListNode]
        ) -> bool:
            if not current_node:
                return True

            if not checkPalindromeRecursively(current_node.next):
                return False

            nonlocal left_node

            result = ((current_node.val) == (left_node.val))

            left_node = left_node.next

            return result

        return checkPalindromeRecursively(head)
